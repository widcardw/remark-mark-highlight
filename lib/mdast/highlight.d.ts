import type {Parent} from 'mdast'

export {markFromMarkdown, markToMarkdown} from './lib.js'

/**
 * Math (text).
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export interface InlineMark extends Parent {
  /**
   * Node type.
   */
  type: 'mark'
  children: PhrasingContent[];
}

// Add custom data tracked to turn a tree into markdown.
declare module 'mdast-util-to-markdown' {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface ConstructNameMap {
    /**
     * Math (flow) meta flag.
     *
     * ```markdown
     * ==a==
     * ^^^^^
     * ```
     */
    markhighlight: 'markhighlight'
  }
}

// Add nodes to tree.
declare module 'mdast' {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface StaticPhrasingContentMap {
    inlineMark: InlineMark
  }
}