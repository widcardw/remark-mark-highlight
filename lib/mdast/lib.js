/**
 * @typedef {import('mdast').Delete} Delete
 *
 * @typedef {import('mdast-util-from-markdown').CompileContext} CompileContext
 * @typedef {import('mdast-util-from-markdown').Extension} FromMarkdownExtension
 * @typedef {import('mdast-util-from-markdown').Handle} FromMarkdownHandle
 *
 * @typedef {import('mdast-util-to-markdown').ConstructName} ConstructName
 * @typedef {import('mdast-util-to-markdown').Options} ToMarkdownExtension
 * @typedef {import('mdast-util-to-markdown').Handle} ToMarkdownHandle
 */

import { containerPhrasing } from 'mdast-util-to-markdown/lib/util/container-phrasing.js'
import { track } from 'mdast-util-to-markdown/lib/util/track.js'

// To do: next major: expose functions.
// To do: next major: use `state`, state utilities.

/**
 * List of constructs that occur in phrasing (paragraphs, headings), but cannot
 * contain strikethrough.
 * So they sort of cancel each other out.
 * Note: could use a better name.
 *
 * Note: keep in sync with: <https://github.com/syntax-tree/mdast-util-to-markdown/blob/8ce8dbf/lib/unsafe.js#L14>
 *
 * @type {Array<ConstructName>}
 */
const constructsWithoutMark = [
  'autolink',
  'destinationLiteral',
  'destinationRaw',
  'reference',
  'titleQuote',
  'titleApostrophe',
]

handleMark.peek = peekMark

/**
 * Extension for `mdast-util-from-markdown` to enable GFM strikethrough.
 *
 * @type {FromMarkdownExtension}
 */
export const markFromMarkdown = {
  canContainEols: ['mark'],
  enter: { markhighlight: enterMarkhighlight },
  exit: { markhighlight: exitMarkhighlight },
}

/**
 * Extension for `mdast-util-to-markdown` to enable GFM strikethrough.
 *
 * @type {ToMarkdownExtension}
 */
export const markToMarkdown = {
  unsafe: [
    {
      character: '=',
      inConstruct: 'phrasing',
      notInConstruct: constructsWithoutMark,
    },
  ],
  handlers: { emphasis: handleMark },
}

/**
 * @this {CompileContext}
 * @type {FromMarkdownHandle}
 */
function enterMarkhighlight(token) {
  this.enter({ type: 'emphasis', children: [] }, token)
}

/**
 * @this {CompileContext}
 * @type {FromMarkdownHandle}
 */
function exitMarkhighlight(token) {
  this.exit(token)
}

/**
 * @type {ToMarkdownHandle}
 * @param {Delete} node
 */
function handleMark(node, _, context, safeOptions) {
  // console.log(node)
  const tracker = track(safeOptions)
  const exit = context.enter('emphasis')
  let value = tracker.move('==')
  value += containerPhrasing(node, context, {
    ...tracker.current(),
    before: value,
    after: '=',
  })
  value += tracker.move('==')
  exit()
  return value
}

/** @type {ToMarkdownHandle} */
function peekMark() {
  return '='
}
