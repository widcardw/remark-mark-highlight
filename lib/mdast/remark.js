/**
 * @typedef {import('mdast').Root} Root
 */

import { microMark } from '../micromark/syntax'
import { markFromMarkdown, markToMarkdown } from './lib'

/**
 * @this {import('unified').Processor}
 * @type {import('unified').Plugin<Root>}
 */
export function remarkMark() {
  const data = this.data()

  add('micromarkExtensions', microMark())
  add('fromMarkdownExtensions', markFromMarkdown)
  add('toMarkdownExtensions', markToMarkdown)

  /**
   * @param {string} field
   * @param {unknown} value
   */
  function add(field, value) {
    const list = /** @type {unknown[]} */ (
      // Other extensions
      /* c8 ignore next 2 */
      data[field] ? data[field] : (data[field] = [])
    )

    list.push(value)
  }
}