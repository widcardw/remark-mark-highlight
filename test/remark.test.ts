import { unified } from 'unified'
import { describe, it, expect } from 'vitest'
import { remarkMark } from '../lib/mdast/remark'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import rehypeStringify from 'rehype-stringify'
import { removePosition } from 'unist-util-remove-position'


const md1 = 'a ==*b*=='

describe('remark', () => {
  it('should generate ast', () => {
    expect(
      removePosition(
        unified()
          .use(remarkParse)
          .use(remarkMark as any)
          .parse(md1),
        true))
      .toMatchInlineSnapshot(`
        {
          "children": [
            {
              "children": [
                {
                  "type": "text",
                  "value": "a ",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "type": "text",
                          "value": "b",
                        },
                      ],
                      "type": "emphasis",
                    },
                  ],
                  "data": {
                    "hName": "mark",
                  },
                  "type": "mark",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "root",
        }
      `)
  })
  it('should parse markdown', () => {
    const toHtml = unified()
      .use(remarkParse)
      .use(remarkMark as any)
      .use(remarkRehype)
      .use(rehypeStringify)

    expect(toHtml.processSync(md1).toString()).toMatchInlineSnapshot('"<p>a <mark><em>b</em></mark></p>"')
  })
})
