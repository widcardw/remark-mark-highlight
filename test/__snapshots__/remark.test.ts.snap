// Vitest Snapshot v1

exports[`remark > should parse markdown 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "a ",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 6,
                  "line": 1,
                  "offset": 5,
                },
                "start": {
                  "column": 5,
                  "line": 1,
                  "offset": 4,
                },
              },
              "type": "text",
              "value": "b",
            },
          ],
          "position": {
            "end": {
              "column": 8,
              "line": 1,
              "offset": 7,
            },
            "start": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "emphasis",
        },
      ],
      "position": {
        "end": {
          "column": 8,
          "line": 1,
          "offset": 7,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
